Dbz 137 implement log miner emitter

AndreyIg opened the pull request:
Hi, we collected so many changes and we have to sync them somehow.
This code we are currently testing, including performance issues.
Starting from now Danny and myself will work of my fork until you merge the code.
This version does Log Mining without CONTINUOUS_MINE option.
We are mining archived logs on the startup, starting from the offset SCN. After catching up, we are mining online redos only.
We believe this mitigates CONTINUOUS_MINE deprecation.
We did not cover dispatching of DDL and some complicated cases like a DDL during downtime, we will address it later. We did not cover all data types yet.

gunnarmorling commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java:
```diff
@@ -0,0 +1,202 @@
+                LOGGER.debug("\nstart, lastProcessedScn: " + lastProcessedScn + ", endScn: " + currentScn);
+                if (lastProcessedScn == currentScn) {
+                    LOGGER.debug("No changes to process...pausing for a second");
+                    Metronome metronome = Metronome.sleeper(Duration.ofSeconds(1L), clock);
```
This should use `CommonConnectorConfig.getPollInterval()`. And the metronome should be set up once and stored in a field.

gunnarmorling commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java:
```diff
@@ -0,0 +1,202 @@
+
+                res.close();
+            }
+        } catch (Throwable e) {
```
Is there any specific need for catching `Throwable`? If not, I'd just let it be passed through.

gunnarmorling commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java:
```diff
@@ -0,0 +1,202 @@
+            try (Connection connection = jdbcConnection.connection()) {
+                LogMinerHelper.endMining(connection);
+            } catch (SQLException e) {
+                LOGGER.error("Cannot close Log Miner session gracefully: " + e.getMessage());
```
Please keep the original exception:
```suggestion
                LOGGER.error("Cannot close Log Miner session gracefully", e);
```

gunnarmorling commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java:
```diff
@@ -0,0 +1,202 @@
+
+    private void dispatchChanges(ResultSet res) throws SQLException {
+
+        LOGGER.debug("dispatching changes..");
```
`trace()` probably?

gunnarmorling commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerStreamingChangeEventSource.java:
```diff
@@ -0,0 +1,202 @@
+                        new ChangeRecordEmitter(offsetContext, logMinerRowLcr, schema.tableFor(tableId), clock)
+                );
+            } catch (InterruptedException e) {
+                LOGGER.error("Following SCN: " + logMinerRowLcr.getActualScn() + " cannot be dispatched due to the : " + e);
```
`InterruptedException` should rather be let passed up to the caller which handles it appropriately.

gunnarmorling commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/OracleChangeRecordValueConverter.java:
```diff
@@ -0,0 +1,501 @@
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.oracle.logminer;
```
Why is this under `logminer`. Isn't it generally used?

gunnarmorling commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/valueholder/LogMinerLcr.java:
```diff
@@ -0,0 +1,84 @@
+     * Sets table name
+     * @param name table name
+     */
+    void setObjectName(String name);
```
Same question as above about mutability.

gunnarmorling commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/ChangeRecordEmitter.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    protected Object[] getOldColumnValues() {
+        List<LogMinerColumnValue> valueList =  lcr.getOldValues();
+        LogMinerColumnValue[] result = Arrays.copyOf(valueList.toArray(), valueList.size(), LogMinerColumnValue[].class );
```
Can we have `LogMinerRowLcr` return arrays from the get-go, then we'd avoid the copying in the hot code path here.

AndreyIg commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/OracleSchemaChangeEventEmitter.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * {@link SchemaChangeEventEmitter} implementation based on Oracle LogMiner utility.
+ */
+public class OracleSchemaChangeEventEmitter extends BaseOracleSchemaChangeEventEmitter {
```
done, will be in the next PR

AndreyIg commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/valueholder/LogMinerRowLcr.java:
```diff
@@ -0,0 +1,59 @@
+     * sets commit scn obtained from a Log Miner entry
+     * @param actualCommitScn the value
+     */
+    void setActualCommitScn(long actualCommitScn);
```
we incorporate SCN info into this object to update offset SCN

AndreyIg commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/valueholder/LogMinerRowLcr.java:
```diff
@@ -0,0 +1,59 @@
+     * the commit scn obtained from a Log Miner entry
+     * @return it's value
+     */
+    long getActualCommitScn();
```
how about BigDecimal?

AndreyIg commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/valueholder/LogMinerRowLcrImpl.java:
```diff
@@ -0,0 +1,152 @@
+        this.transactionId = id;
+    }
+
+    public void setSourceDatabaseName(String name){
```
yes, we can do it, we can save database name in the parser and construct objects with this name.
yes, we are not changing database dynamically

AndreyIg commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/xstream/OracleSchemaChangeEventEmitter.java:
```diff
@@ -0,0 +1,29 @@
+ *
+ * @author Gunnar Morling
+ */
+public class OracleSchemaChangeEventEmitter extends BaseOracleSchemaChangeEventEmitter {
```
we did

AndreyIg commented on debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/LogMinerOracleConnectorIT.java:
```diff
@@ -0,0 +1,100 @@
+ * This subclasses common OracleConnectorIT for LogMiner adaptor
+ *
+ */
+public class LogMinerOracleConnectorIT extends OracleConnectorIT {
```
we will review it

AndreyIg commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerProcessor.java:
```diff
@@ -0,0 +1,121 @@
+            }
+
+            if(!tableParsed){
+                LOGGER.debug("table " + tableName + " is not in the whitelist, skipped");
```
sure

AndreyIg commented on debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerProcessor.java:
```diff
@@ -0,0 +1,121 @@
+            }
+
+            //TCL
+            if (operationCode == 7 || operationCode == 36) {
```
we were planning to do this plus all other constants

AndreyIg commented:
we are currently addressing one serious problem: some changes are missed due to the REDO LOG flush timing.
When we address it, we will return to all those above

AndreyIg commented:
We got in sync with Dany again. In this commit we addressed 
- a few community comments,
- implemented sdo_geometry, CLOB and BLOB simple conversion to string,
- make it working with no container database
- fixed timestamp problems,
- handled redo log files switches, which require rebuilding dictionary and readding files
- parsing table alias
- replaced metadata usage with sql queries for couple of cases, still have to address columns metadata
- handled UNUSED state of redo log files

Daniel-Yosief commented:
Added TransactionalBuffer concept from @jchipmunk.

Daniel-Yosief commented:
Fix logic to update SCN on offset context

gunnarmorling commented:
Hey @Daniel-Yosief and @AndreyIg, I'm back from three weeks of vacation :) How are things for you? Are you planning to more work to this right now? Or should I do another review pass? Thanks!

gunnarmorling commented:
Btw. could you rebase this PR to the latest master? It has some conflicts.

Also, this blog post could be interesting to read: https://medium.com/@FranckPachot/ideas-for-event-sourcing-in-oracle-d4e016e90af6. It discusses some CDC options for Oracle, but seems to arrive at the same conclusion as you do: use LogMiner without "continuous mine".

gunnarmorling commented:
@AndreyIg, @Daniel-Yosief, howdy? Should we sync on this PR again?

kpmeen commented:
Has there been any (hidden) progress on this PR? As far as I can tell nothing has happened in the last 4 months or so.

gunnarmorling commented:
@kpmeen, yes, @AndreyIg is working on this; I'll refer to him for the details, but this PR should be updated within the next few weeks.

kpmeen commented:
That's great new @gunnarmorling . Do you have any (rough) estimates for when this might be merged to master and made available in a release (or RC)? If you need, I'd love to help testing it to speed things up.

jchipmunk commented:
@Daniel-Yosief, @AndreyIg, I read your MR and I have the following questions:
1. Why aren’t you mine data from archive logs?
2. How do you resolve the problem of cyclic overwriting of online log files?

>LogMiner can mine online redo logs. However, if the CONTINUOUS_MINE option is not specified, it is possible that the database is writing to the online redo log file at the same time that LogMiner is reading the online redo log file. If a log switch occurs while LogMiner is reading an online redo log file, the database will overwrite what LogMiner is attempting to read. The data that LogMiner returns if the file it is trying to read gets overwritten by the database is unpredictable.

For more information, see [DBMS_LOGMNR](https://docs.oracle.com/database/121/ARPLS/d_logmnr.htm#ARPLS66809).

AndreyIg commented:
Hi, Andrey
1. We are currently OK to mine up to 1 hour back in history. We assume the
connector will not be down for a longer time. We require 7 redo groups 0.5G
each file.
For our load it is sufficient (15min switch * 7)
Since continious_mine is deprecated (and throws on Oracle 19C which is
supposed to be the longest supported version) we add/remove files manually.
For the first release we want to have a simple algorithm as much as
possible. No archives, no DDL
2. When next switch has happen, it turns state to "active" to archive it's
content. It does not get overwritten until end of full cycle (7 switches in
this case). The file changes its states: current --> active -> inactive.
The mining gets paused when file is in active state (a few seconds).

We are in testing phase and already passed many scenarios. Currently we are
addressing PGA leak on windows installation.
I included Andrew - the third Andrey :)  who is our DBA. Please correct us
if we are wrong.

Thank you,





On Tue, Dec 10, 2019 at 5:10 AM Andrey Pustovetov <notifications@github.com>
wrote:

> @Daniel-Yosief <https://github.com/Daniel-Yosief>, @AndreyIg
> <https://github.com/AndreyIg>, I read your MR and I have the following
> questions:
>
>    1. Why aren’t you mine data from archive logs?
>    2. How do you resolve the problem of cyclic overwriting of online log
>    files?
>
> LogMiner can mine online redo logs. However, if the CONTINUOUS_MINE option
> is not specified, it is possible that the database is writing to the online
> redo log file at the same time that LogMiner is reading the online redo log
> file. If a log switch occurs while LogMiner is reading an online redo log
> file, the database will overwrite what LogMiner is attempting to read. The
> data that LogMiner returns if the file it is trying to read gets
> overwritten by the database is unpredictable.
>
> For more information, see DBMS_LOGMNR
> <https://docs.oracle.com/database/121/ARPLS/d_logmnr.htm#ARPLS66809>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/debezium/debezium-incubator/pull/96?email_source=notifications&email_token=AL3FDR2U3XNWBC2RLNWHIODQX6IMFA5CNFSM4HY3MPIKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGPFOTA#issuecomment-564025164>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AL3FDRYRCVWUBMOQXBYVUIDQX6IMFANCNFSM4HY3MPIA>
> .
>

kpmeen commented:
Hello @AndreyIg, thanks for your effort on this! I'm looking forward to try this connector. Do you have any rough estimates on how far away you are from being able to merge and release? Will there be a beta or pre-release that I could try out?

AndreyIg commented:
Hi, this version is in pre-production testing phase.
I would appreciate if you can try and test it too.
Currently I'm waiting Gunnar Morling's review, suggestions and help.
I have to capacity to merge it soon.
Thank you,
Andrey

On Sun, Jan 19, 2020 at 11:53 PM Knut Petter Meen <notifications@github.com>
wrote:

> Hello @AndreyIg <https://github.com/AndreyIg>, thanks for your effort on
> this! I'm looking forward to try this connector. Do you have any rough
> estimates on how far away you are from being able to merge and release?
> Will there be a beta or pre-release that I could try out?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/debezium/debezium-incubator/pull/96?email_source=notifications&email_token=AL3FDR76J6E3AR32E6AREQLQ6VKAPA5CNFSM4HY3MPIKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJLVTOY#issuecomment-576149947>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AL3FDRY3BYTYFTLVHGOLQRLQ6VKAPANCNFSM4HY3MPIA>
> .
>

thake commented:
I just now noticed this pull request. In the last weeks I've also written a Kafka Connector with logminer support (https://github.com/thake/logminer-kafka-connect). As your attempt seems to be more mature, I would be happy to contribute.

I noticed that you are just reading committed records. How do you cope with very large transactions? Logminer might just give up due to memory constraints. Also queries against V$LOGMNR_CONTENTS might take significantly longer.

> If the COMMITTED_DATA_ONLY option is specified and you issue a query, then LogMiner stages all redo records within a single transaction in memory until LogMiner finds the commit record for that transaction. Therefore, it is possible to exhaust memory, in which case an "Out of Memory" error will be returned. If this occurs, then you must restart LogMiner without the COMMITTED_DATA_ONLY option specified and reissue the query. (https://docs.oracle.com/database/121/SUTIL/GUID-6A2398F7-D484-495A-8AD2-0A6B34C03536.htm#SUTIL1574)

Because of this I switched to reading uncommitted data. Uncommitted records will be stored to a memory mapped file and will be only processed if a commit occurs. Have a look at https://github.com/thake/logminer-kafka-connect/blob/master/src/main/kotlin/com/github/thake/logminer/kafka/connect/logminer/Transaction.kt.

FrankMormino commented:
@thake @AndreyIg - awesome work here -  thanks for your effort on this! I'm looking forward to try this connector. Do you have any rough estimates on how far away you are from being able to merge and release? Will there be a beta or pre-release that I could try out?

AndreyIg commented:
Sorry for the late reply.
I'm really busy.
The connector is passing last stages of scalability and resiliency testing.
I'll update my for as soon as possible. I hope next week.
Thanks.


On Sun, Jan 19, 2020 at 11:53 PM Knut Petter Meen <notifications@github.com>
wrote:

> Hello @AndreyIg <https://github.com/AndreyIg>, thanks for your effort on
> this! I'm looking forward to try this connector. Do you have any rough
> estimates on how far away you are from being able to merge and release?
> Will there be a beta or pre-release that I could try out?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/debezium/debezium-incubator/pull/96?email_source=notifications&email_token=AL3FDR76J6E3AR32E6AREQLQ6VKAPA5CNFSM4HY3MPIKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJLVTOY#issuecomment-576149947>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AL3FDRY3BYTYFTLVHGOLQRLQ6VKAPANCNFSM4HY3MPIA>
> .
>

FrankMormino commented:
@thake @AndreyIg - do you have any notes for anyone who wants to test this out for you? I have a scenario where we could utilize this and I'm trying to get this to work against oracle logminer instance 11.2.0.3 - I have spent day or two navigating the documentation from debezium to debezium-incubator and want to make sure I am configuring and building this connector correctly. Is there anywhere you can point me for this or GOTCHAs you've ran into while you tested this out?

crognon commented:
Hi,

Thank you for the good job ! how can we test your work to report feedback here ?

thanks
Cyril

AndreyIg commented:
Hi, We are planning to merge it to the Debezuim base this or maybe next
week.

On Tue, Jun 23, 2020 at 6:02 AM crognon <notifications@github.com> wrote:

> Hi,
>
> Thank you for the good job ! how can we test your work to report feedback
> here ?
>
> thanks
> Cyril
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/debezium/debezium-incubator/pull/96#issuecomment-648130722>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AL3FDR5MT36I2ZEZIGFBDFTRYCRWLANCNFSM4HY3MPIA>
> .
>

Naros commented:
Superseded by PR #185.
